src/Internal/BalanceTx/Sync.purs

```diff
@@ -0,0 +1,275 @@
+  { delay: delayMs, timeout } <- asks $ _.timeParams >>> _.syncBackend
+  { errorOnTimeout } <- asks $
+    _.synchronizationParams >>> _.syncBackendWithWallet
+  let
+    errorMessage = "Failed to wait for wallet state synchronization (timeout). "
+      <> guard (not errorOnTimeout) "Continuing anyway. "
+      <> "Consider increasing `timeParams.syncBackend.timeout`"
+      <> " in `ContractParams`"
+  knownTxsRef <- asks (_.knownTxs >>> _.backend)
+  let
+    sync :: Contract Unit
+    sync = do
+      utxos <- getControlledUtxos
+      alreadyFoundTxs <- liftEffect $ Ref.read knownTxsRef
+      let
+        isAlreadyFoundUtxo =
+          flip Set.member alreadyFoundTxs <<< _.transactionId <<< unwrap
+
+        -- all utxos from the wallet, except of those we have already seen
+        notYetFoundUtxos :: Array TransactionInput
+        notYetFoundUtxos =
+          Array.filter (not <<< isAlreadyFoundUtxo)
+            $ fst <$> Map.toUnfoldableUnordered utxos
+      -- split the UTxOs into known to the query layer and not yet known
+      { yes: newlyFound, no: stillNotFound } <- Array.partition (snd >>> isJust)
+        <$> parTraverse (\input -> Tuple input <$> getUtxo' input)
+          notYetFoundUtxos
+      -- Mark newly found TxIds as known
+      liftEffect $ Ref.modify_
+        ( Set.union
+            $ Set.fromFoldable
+            $ newlyFound <#> fst >>> unwrap >>> _.transactionId
+        )
+        knownTxsRef
+      if Array.null stillNotFound then do
+        logTrace' "syncBackendWithWallet: synchronization finished"
+      else do
+        logTrace' $
+          "syncBackendWithWallet: waiting for query layer state " <>
+            "synchronization with the wallet..."
+        liftAff (delay delayMs)
+        sync
+  parOneOf
+    [ sync
+    , waitAndError errorOnTimeout (fromDuration timeout) errorMessage
+    ]
+
+-- | Wait until the wallet can see the UTxOs created by a given transaction.
+-- | This function assumes that the transaction has already been confirmed
+-- | (i.e. the backend query layer can see its UTxOs).
+-- | This function does not have any effect if `KeyWallet` is used.
+-- | This function does not have any effect if none of the outputs are at
+-- | addresses controlled by the wallet.
+-- | The assumption is that the transaction has at least one output and that it
+-- | hasn't been consumed by another transaction yet.
+-- | You don't need to use this function if you use `awaitTxConfirmed*`.
+syncWalletWithTransaction :: TransactionHash -> Contract Unit
+syncWalletWithTransaction txHash = whenM isCip30Wallet do
+  { delay: delayMs, timeout } <- asks (_.timeParams >>> _.syncWallet)
+  { errorOnTimeout } <- asks $
+    _.synchronizationParams >>> _.syncWalletWithTransaction
+  let
+    errorMessage = "syncWalletWithTransaction: Failed to wait for wallet state "
+      <> "synchronization. "
+      <> guard (not errorOnTimeout) "Continuing anyway. "
+      <> "This may indicate UTxO locking"
```

Comment:
Might be worth mentioning the explanation about why  UTxO locking is problematic here or in the docstring